src/compute/src/render/join/mz_join_core.rs

```diff
@@ -631,10 +630,12 @@ where
                             if let Some(first) = result.next() {
                                 // Join times.
                                 cursor1.map_times(storage1, |time1, diff1| {
-                                    let time1 = time1.join(meet);
+                                    let mut time1 = time1.into_owned();
+                                    time1.join_assign(meet);
+                                    let diff1 = diff1.into_owned();
                                     cursor2.map_times(storage2, |time2, diff2| {
-                                        let time = time1.join(time2);
-                                        let diff = diff1.multiply(diff2);
+                                        let time = time1.join(&time2.into_owned());
+                                        let diff = diff1.multiply(&diff2.into_owned());
```

Comment:
I think these could have fewer allocations, e.g. taking `time2.into_owned` and `join_assign`ing into it.